### PR TITLE
[core] Support breakpoints

### LINKED ISSUE #16: BOOM lacks breakpoint support
BOOM currently lacks breakpoint support.  It would be nice to have it.

Things to note: 

1. The boom branch of rocket-chip disables the `breakpoint` test (so the necessary software to test this feature already exists):

https://github.com/ucb-bar/rocket-chip/blob/boom/src/main/scala/rocketchip/RocketTestSuite.scala#L153

2. The rocket core implements breakpoints and provides a lot of starting pointers for a BOOM implementation:

https://github.com/ucb-bar/rocket-chip/blob/master/src/main/scala/rocket/breakpoint.scala 
https://github.com/ucb-bar/rocket-chip/blob/master/src/main/scala/rocket/rocket.scala#L268

3. One of the challenges is handling a breakpoint on an instruction address in the presence of superscalar fetching.

### PR DESCRIPTION
**Pull request by jerryz123**
**Related issue**: Resolves #16 

**Type of change**: new feature

**Impact**:  new rtl

**Development Phase**: implementation

**Release Notes**

Breakpoints now supported (I think). We can use Rocket's `DefaultTestSuite` now.

### CODE REVIEW

**Comment by jerryz123** [comment]
I should probably double-check this doesn't break buildroot/Fedora.

File: src/main/scala/ifu/fetch-control-unit.scala [hunk 1]
```diff
@@ -28,6 +28,7 @@ import chisel3.util._
 import chisel3.core.{withReset, DontCare}
 import chisel3.experimental.{dontTouch}
 
+import freechips.rocketchip.rocket.{MStatus, BP, BreakpointUnit}
```
Comment (bkorpan) [inline src/main/scala/ifu/fetch-control-unit.scala]: The BreakpointUnit isn't used in this file.

**Review by bkorpan — COMMENTED** [review]
Potential for some QoR impact, but probably alright. I'm not a fan of instantiating a BreakpointUnit in every pipeline slot of the fetch buffer enqueue logic, it seems like this could be done more efficiently, though it might be nasty code.

File: src/main/scala/ifu/fetch-control-unit.scala (hunk 1 above)
Comment (jerryz123) [reply src/main/scala/ifu/fetch-control-unit.scala]: Good catch.

**Comment by jerryz123** [comment]
I defaulted `nBreakpoints` to 0, which should make the BreakPointUnits dummy modules that always output `false`.
When this gets merged, I'll open an issue that mentions updating the frontend breakpointunit.

**Review by bkorpan — APPROVED** [review]

**Review by abejgonzalez — APPROVED** [review]
LGTM as long as the default behavior is BP = 0 for now.

**Comment by abejgonzalez** [comment]
Lets wait on this PR until the Chipyard PR gets merged in, that will remove the conflict on the BoomTestSuites file

**Comment by jerryz123** [comment]
Fine with me

**Comment by abejgonzalez** [comment]
Why does this still conflict? Did the rebase work correctly?

**Comment by jerryz123** [comment]
> Why does this still conflict? Did the rebase work correctly?

Looks like I messed up. I'll rebase again.

**Pull request merged**